File: javascript/packages/printer/src/indent-printer.ts
import { IdentityPrinter } from "./identity-printer.js"

import type * as Nodes from "@herb-tools/core"

/**
 * IndentPrinter - Re-indentation printer that preserves content but adjusts indentation
 *
 * Extends IdentityPrinter to preserve all content as-is while replacing
 * leading whitespace on each line with the correct indentation based on
 * the AST nesting depth.
 */
export class IndentPrinter extends IdentityPrinter {
  protected indentLevel: number = 0
  protected indentWidth: number
  private pendingIndent: boolean = false

  constructor(indentWidth: number = 2) {
    super()

    this.indentWidth = indentWidth
  }

  protected get indent(): string {
    return " ".repeat(this.indentLevel * this.indentWidth)
  }

  protected write(content: string): void {
    if (this.pendingIndent && content.length > 0) {
      this.pendingIndent = false
      this.context.write(this.indent + content)
    } else {
      this.context.write(content)
    }
  }

  visitLiteralNode(node: Nodes.LiteralNode): void {
    this.writeWithIndent(node.content)
  }

  visitHTMLTextNode(node: Nodes.HTMLTextNode): void {
    this.writeWithIndent(node.content)
  }

  visitHTMLElementNode(node: Nodes.HTMLElementNode): void {
    const tagName = node.tag_name?.value

    if (tagName) {
      this.context.enterTag(tagName)
    }

    if (node.open_tag) {
      this.visit(node.open_tag)
    }

    if (node.body) {
      this.indentLevel++
      node.body.forEach(child => this.visit(child))
      this.indentLevel--
    }

    if (node.close_tag) {
      this.visit(node.close_tag)
    }

    if (tagName) {
      this.context.exitTag()
    }
  }

  visitERBIfNode(node: Nodes.ERBIfNode): void {
    this.printERBNode(node)

    if (node.statements) {
      this.indentLevel++
      node.statements.forEach(statement => this.visit(statement))
      this.indentLevel--
    }

    if (node.subsequent) {
      this.visit(node.subsequent)
    }

    if (node.end_node) {
      this.visit(node.end_node)
    }
  }

  visitERBElseNode(node: Nodes.ERBElseNode): void {
    this.printERBNode(node)

    if (node.statements) {
      this.indentLevel++
      node.statements.forEach(statement => this.visit(statement))
      this.indentLevel--
    }
  }

  visitERBBlockNode(node: Nodes.ERBBlockNode): void {
    this.printERBNode(node)

    if (node.body) {
      this.indentLevel++
      node.body.forEach(child => this.visit(child))
      this.indentLevel--
    }

    if (node.rescue_clause) {
      this.visit(node.rescue_clause)
    }

    if (node.else_clause) {
      this.visit(node.else_clause)
    }

    if (node.ensure_clause) {
      this.visit(node.ensure_clause)
    }

    if (node.end_node) {
      this.visit(node.end_node)
    }
  }

  visitERBCaseNode(node: Nodes.ERBCaseNode): void {
    this.printERBNode(node)

    if (node.children) {
      this.indentLevel++
      node.children.forEach(child => this.visit(child))
      this.indentLevel--
    }

    if (node.conditions) {
      this.indentLevel++
      node.conditions.forEach(condition => this.visit(condition))
      this.indentLevel--
    }

    if (node.else_clause) {
      this.indentLevel++
      this.visit(node.else_clause)
      this.indentLevel--
    }

    if (node.end_node) {
      this.visit(node.end_node)
    }
  }

  visitERBWhenNode(node: Nodes.ERBWhenNode): void {
    this.printERBNode(node)

    if (node.statements) {
      this.indentLevel++
      node.statements.forEach(statement => this.visit(statement))
      this.indentLevel--
    }
  }

  visitERBWhileNode(node: Nodes.ERBWhileNode): void {
    this.printERBNode(node)

    if (node.statements) {
      this.indentLevel++
      node.statements.forEach(statement => this.visit(statement))
      this.indentLevel--
    }

    if (node.end_node) {
      this.visit(node.end_node)
    }
  }

  visitERBUntilNode(node: Nodes.ERBUntilNode): void {
    this.printERBNode(node)

    if (node.statements) {
      this.indentLevel++
      node.statements.forEach(statement => this.visit(statement))
      this.indentLevel--
    }

    if (node.end_node) {
      this.visit(node.end_node)
    }
  }

  visitERBForNode(node: Nodes.ERBForNode): void {
    this.printERBNode(node)

    if (node.statements) {
      this.indentLevel++
      node.statements.forEach(statement => this.visit(statement))
      this.indentLevel--
    }

    if (node.end_node) {
      this.visit(node.end_node)
    }
  }

  visitERBBeginNode(node: Nodes.ERBBeginNode): void {
    this.printERBNode(node)

    if (node.statements) {
      this.indentLevel++
      node.statements.forEach(statement => this.visit(statement))
      this.indentLevel--
    }

    if (node.rescue_clause) {
      this.visit(node.rescue_clause)
    }

    if (node.else_clause) {
      this.visit(node.else_clause)
    }

    if (node.ensure_clause) {
      this.visit(node.ensure_clause)
    }

    if (node.end_node) {
      this.visit(node.end_node)
    }
  }

  visitERBRescueNode(node: Nodes.ERBRescueNode): void {
    this.printERBNode(node)

    if (node.statements) {
      this.indentLevel++
      node.statements.forEach(statement => this.visit(statement))
      this.indentLevel--
    }

    if (node.subsequent) {
      this.visit(node.subsequent)
    }
  }

  visitERBEnsureNode(node: Nodes.ERBEnsureNode): void {
    this.printERBNode(node)

    if (node.statements) {
      this.indentLevel++
      node.statements.forEach(statement => this.visit(statement))
      this.indentLevel--
    }
  }

  visitERBUnlessNode(node: Nodes.ERBUnlessNode): void {
    this.printERBNode(node)

    if (node.statements) {
      this.indentLevel++
      node.statements.forEach(statement => this.visit(statement))
      this.indentLevel--
    }

    if (node.else_clause) {
      this.visit(node.else_clause)
    }

    if (node.end_node) {
      this.visit(node.end_node)
    }
  }

  /**
   * Write content, replacing leading whitespace on each line with the current indent.
   *
   * Uses a pendingIndent mechanism: when content ends with a newline followed by
   * whitespace-only, sets pendingIndent=true instead of writing the indent immediately.
   * The indent is then applied at the correct level when the next node writes content
   * (via the overridden write() method).
   */
  protected writeWithIndent(content: string): void {
    if (!content.includes("\n")) {
      if (this.pendingIndent) {
        this.pendingIndent = false

        const trimmed = content.replace(/^[ \t]+/, "")

        if (trimmed.length > 0) {
          this.context.write(this.indent + trimmed)
        }
      } else {
        this.context.write(content)
      }

      return
    }

    const lines = content.split("\n")
    const lastIndex = lines.length - 1

    for (let i = 0; i < lines.length; i++) {
      if (i > 0) {
        this.context.write("\n")
      }

      const line = lines[i]
      const trimmed = line.replace(/^[ \t]+/, "")

      if (i === 0) {
        if (this.pendingIndent) {
          this.pendingIndent = false

          if (trimmed.length > 0) {
            this.context.write(this.indent + trimmed)
          }
        } else {
          this.context.write(line)
        }
      } else if (i === lastIndex && trimmed.length === 0) {
        this.pendingIndent = true
      } else if (trimmed.length === 0) {
        // Middle whitespace-only line: write nothing (newline already written above)
      } else {
        this.context.write(this.indent + trimmed)
      }
    }
  }
}

File: rust/herb-printer/src/indent_printer.rs
use herb::nodes::DocumentNode;
use herb::Visitor;

use crate::print_context::PrintContext;
use crate::printer::Printer;
use crate::printing_visitor::impl_printing_visitor;

pub struct IndentPrinter {
  context: PrintContext,
  indent_level: usize,
  indent_width: usize,
  pending_indent: bool,
}

impl Default for IndentPrinter {
  fn default() -> Self {
    Self::new(2)
  }
}

impl IndentPrinter {
  pub fn new(indent_width: usize) -> Self {
    Self {
      context: PrintContext::new(),
      indent_level: 0,
      indent_width,
      pending_indent: false,
    }
  }

  pub fn print_document(document: &DocumentNode) -> String {
    Self::print_with_width(document, 2)
  }

  pub fn print_with_width(document: &DocumentNode, indent_width: usize) -> String {
    let mut printer = Self::new(indent_width);

    printer.visit_document_node(document);

    printer.context.output().to_string()
  }

  fn indentation(&self) -> String {
    " ".repeat(self.indent_level * self.indent_width)
  }
}

impl Printer for IndentPrinter {
  fn context(&mut self) -> &mut PrintContext {
    &mut self.context
  }

  fn context_ref(&self) -> &PrintContext {
    &self.context
  }

  fn enter_indent(&mut self) {
    self.indent_level += 1;
  }

  fn exit_indent(&mut self) {
    self.indent_level = self.indent_level.saturating_sub(1);
  }

  fn write(&mut self, content: &str) {
    if self.pending_indent && !content.is_empty() {
      self.pending_indent = false;

      let indentation = self.indentation();
      self.context.write(&indentation);
      self.context.write(content);
    } else {
      self.context.write(content);
    }
  }

  fn write_text(&mut self, content: &str) {
    if !content.contains('\n') {
      if self.pending_indent {
        self.pending_indent = false;

        let trimmed = content.trim_start_matches([' ', '\t']);

        if !trimmed.is_empty() {
          let indentation = self.indentation();
          self.context.write(&indentation);
          self.context.write(trimmed);
        }
      } else {
        self.context.write(content);
      }

      return;
    }

    let lines: Vec<&str> = content.split('\n').collect();
    let last_index = lines.len() - 1;

    for (index, line) in lines.iter().enumerate() {
      if index > 0 {
        self.context.write("\n");
      }

      let trimmed = line.trim_start_matches([' ', '\t']);

      if index == 0 {
        if self.pending_indent {
          self.pending_indent = false;

          if !trimmed.is_empty() {
            let indentation = self.indentation();
            self.context.write(&indentation);
            self.context.write(trimmed);
          }
        } else {
          self.context.write(line);
        }
      } else if index == last_index && trimmed.is_empty() {
        self.pending_indent = true;
      } else if trimmed.is_empty() {
        // middle whitespace-only line: the newline above is enough
      } else {
        let indentation = self.indentation();
        self.context.write(&indentation);
        self.context.write(trimmed);
      }
    }
  }
}

impl_printing_visitor!(IndentPrinter);
